datanode/candlesv2/service.go

```diff
@@ -49,54 +48,51 @@ func NewService(ctx context.Context, log *logging.Logger, config Config, candleS
 	log.SetLevel(config.Level.Get())
 
 	return &Svc{
-		ctx:                      ctx,
-		log:                      log,
-		Config:                   config,
-		CandleStore:              candleStore,
-		candleIDToUpdatesStream:  map[string]*CandleUpdates{},
-		subscriptionIDToCandleID: map[string]string{},
+		ctx:                     ctx,
+		log:                     log,
+		Config:                  config,
+		CandleStore:             candleStore,
+		candleIDToUpdatesStream: map[string]*CandleUpdates{},
 	}
 }
 
 // Subscribe to a channel of new or updated candles. The subscriber id will be returned as an uint64 value
 // and must be retained for future reference and to Unsubscribe.
-func (cs *Svc) Subscribe(ctx context.Context, candleID string) (string, <-chan entities.Candle, error) {
+func (cs *Svc) Subscribe(ctx context.Context, candleID string) (uint64, <-chan []entities.Candle, error) {
 	cs.updatesSubscriptionMutex.Lock()
 	defer cs.updatesSubscriptionMutex.Unlock()
 
 	exists, err := cs.CandleExists(ctx, candleID)
 	if err != nil {
-		return "", nil, fmt.Errorf("subscribing to candles:%w", err)
+		return 0, nil, fmt.Errorf("subscribing to candles:%w", err)
 	}
 
 	if !exists {
-		return "", nil, fmt.Errorf("no candle exists for candle id:%s", candleID)
+		return 0, nil, fmt.Errorf("no candle exists for candle id:%s", candleID)
 	}
 
-	if _, ok := cs.candleIDToUpdatesStream[candleID]; !ok {
-		updates, err := NewCandleUpdates(cs.ctx, cs.log, candleID, cs, cs.Config.CandleUpdates)
+	updatesStream, ok := cs.candleIDToUpdatesStream[candleID]
+
+	// If we don't have a stream for this candle, or existing update stream has stopped running, make a new one.
+	if !ok || (ok && !updatesStream.IsRunning()) {
```

Comment:
Possiblility here that a subscribing client ends us subscribed to a dead updates stream, as the updates stream dies after this call, but the subscriber thread go on to subscribe thinking all is well, but it will now be subscribed to a dead stream

Reply:
Yes you are right again. Spent an inordinate amount of time trying to come up with the best solution to this and I think I am hopefully there now. There's now a mutex that gets taken when the UpdateStream is in the process of shutting down that prevents new subscriptions.